File: server/src/Connection/Messages/EnvironmentInfo.cs
using System.Text.Json.Serialization;

namespace GameServer.Connection;

public record EnvironmentInfoMessage : Message
{
    [JsonPropertyName("messageType")]
    public override string MessageType { get; init; } = "ENVIRONMENT_INFO";

    [JsonPropertyName("walls")]
    public List<Wall> Walls { get; init; } = new();

    public record Wall
    {
        [JsonPropertyName("position")]
        public Position position { get; init; } = new();

        public record Position
        {
            [JsonPropertyName("x")]
            public double X { get; init; }

            [JsonPropertyName("y")]
            public double Y { get; init; }

            [JsonPropertyName("angle")]
            public double Angle { get; init; }
        }

    }

    [JsonPropertyName("fences")]
    public List<Fence> Fences { get; init; } = new();

    public record Fence
    {
        [JsonPropertyName("position")]
        public Position position { get; init; } = new();

        public record Position
        {
            [JsonPropertyName("x")]
            public double X { get; init; }

            [JsonPropertyName("y")]
            public double Y { get; init; }

            [JsonPropertyName("angle")]
            public double Angle { get; init; }
        }

        [JsonPropertyName("health")]
        public int Health { get; init; }
    }

    [JsonPropertyName("bullets")]
    public List<Bullet> Bullets { get; init; } = new();

    public record Bullet
    {
        [JsonPropertyName("position")]
        public Position Position { get; init; } = new();

        public record Position
        {
            [JsonPropertyName("x")]
            public double X { get; init; }

            [JsonPropertyName("y")]
            public double Y { get; init; }

            [JsonPropertyName("angle")]
            public double Angle { get; init; }
        }

        [JsonPropertyName("speed")]
        public double Speed { get; init; }

        [JsonPropertyName("damage")]
        public double Damage { get; init; }

        [JsonPropertyName("traveledDistance")]
        public double TraveledDistance { get; init; }
    }

    [JsonPropertyName("playerPositions")]
    public List<PlayerPosition> PlayerPositions { get; init; } = new();

    public record PlayerPosition
    {
        [JsonPropertyName("position")]
        public Position Position { get; init; } = new();

        public record Position
        {
            [JsonPropertyName("x")]
            public double X { get; init; }

            [JsonPropertyName("y")]
            public double Y { get; init; }

            [JsonPropertyName("angle")]
            public double Angle { get; init; }
        }

        [JsonPropertyName("token")]
        public string Token { get; init; } = "";
    }
}

File: sdk/python/src/agent/environment_info.py
from typing import List

from .position import Position
from .position_int import PositionInt


class Wall:
    def __init__(self, position: PositionInt):
        self.position = position

    def __str__(self):
        return f"Wall(position={self.position})"


class Fence:
    def __init__(self, position: PositionInt, health: int):
        self.position = position
        self.health = health

    def __str__(self):
        return f"Fence(position={self.position}, health={self.health})"


class Bullet:
    def __init__(
        self,
        no: int,
        isMissile: bool,
        isAntiArmor: bool,
        position: Position,
        speed: float,
        damage: float,
        traveled_distance: float,
    ):
        self.no = no
        self.isMissile = isMissile
        self.isAntiArmor = isAntiArmor
        self.position = position
        self.speed = speed
        self.damage = damage
        self.traveled_distance = traveled_distance

    def __str__(self):
        return f"Bullet(position={self.position}, speed={self.speed}, damage={self.damage}, traveled_distance={self.traveled_distance})"


class EnvironmentInfo:
    def __init__(
        self,
        walls: List[Wall],
        fences: List[Fence],
        bullets: List[Bullet],
        map_size: int,
    ):
        self.walls = walls
        self.fences = fences
        self.bullets = bullets
        self.map_size = map_size

    def __str__(self):
        return f"EnvironmentInfo(walls={self.walls}, fences={self.fences}, bullets={self.bullets}, map_size={self.map_size})"
